### PR TITLE
docs(design): evaluate SARIF ingestion as a shared parser layer

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Commit Summary (Conventional Commits)

- Title: `docs(design): evaluate SARIF ingestion as a shared parser layer`
- Type: **docs** (no version bump)
- Breaking change: no

## What's Changing

Research spike (evaluation, **not** a full implementation) for issue #1066:
should a shared SARIF 2.1.0 ingestion path replace bespoke per-tool output
parsers?

- **`docs/design/sarif-ingestion-evaluation.md`** — the deliverable. Per-tool
  SARIF support matrix, `BaseIssue` model-mapping analysis, LOC delete/add
  estimate, risk table (fix-suggestion / severity / doc-URL fidelity,
  performance, version-gating), and a **partial-adopt** recommendation with a
  5-phase rollout that keeps native parsers as a fallback.
- **`lintro/parsers/sarif/`** — minimal proof-of-concept: `SarifIssue(BaseIssue)`
  and `parse_sarif_output()`. Deliberately **not** registered with any tool
  plugin; **no existing parser or tool definition is modified**.
- **Tests** — `tests/unit/parsers/test_sarif_parser.py` validates the mapping
  against **real** ruff and semgrep SARIF fixtures plus synthetic edge cases
  (severity levels, `ruleIndex` resolution, `file://` decoding, location-free
  results, multi-run logs, malformed input).
- **`pyproject.toml`** — registers the new `lintro.parsers.sarif` package.

## Key findings

- 5 tools emit SARIF natively (verified in-env): **ruff, semgrep, gitleaks,
  hadolint, osv-scanner**. Bandit needs a formatter plugin; clippy/mypy need
  converters; the rest have no native path (formatters/pytest are N/A).
- Core-field mapping (file/line/column/message/code/severity) is clean.
  Fidelity losses: ruff's safe/unsafe fix flag collapses to a boolean; hadolint
  emits **no** `rules[]` so doc URLs are lost; per-tool enrichment (bandit CWE,
  semgrep category) lives in tool-specific `properties`.
- ~324 LOC added (shared parser) could retire ~700–900 LOC of bespoke parsing
  across the 5 native tools — the win is concentration, hardening one parser
  against drift instead of five.

## Checklist

- [x] Title follows Conventional Commits
- [x] Tests added/updated
- [x] Docs updated if user-facing (docs-only spike)
- [x] Local CI passed (`uv run lintro chk`, `uv run lintro fmt`, full `pytest`
      green except two known env-only pre-existing failures)

## Related Issues

Refs #1066

## Details

This is an evaluation spike, not a merge candidate for implementation. Nothing
is wired into lintro's execution path; the PoC exists to validate the mapping.
Full test suite: 6012 passed, 10 skipped (the two known env-only failures —
`test_markdownlint_direct_vs_lintro_parity`,
`test_prepare_execution_version_check_fails_returns_early_result` — deselected).

<details><summary><h3>Greptile Summary</h3></summary>

This PR evaluates SARIF ingestion as a shared parser layer.

- Adds a design document with the SARIF support matrix, mapping analysis, risks, and phased rollout plan.
- Adds a proof-of-concept SARIF parser and issue model.
- Registers the new parser package.
- Adds real ruff and semgrep SARIF fixtures plus parser unit tests.
</details>

<h3>Confidence Score: 5/5</h3>

This looks safe to merge.

- No blocking issues found in the changed code.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| docs/design/sarif-ingestion-evaluation.md | Adds the SARIF evaluation, risk table, and partial-adoption plan. |
| lintro/parsers/sarif/sarif_parser.py | Adds a defensive SARIF parser proof of concept that is not wired into tool execution. |
| lintro/parsers/sarif/sarif_issue.py | Adds the SARIF issue model on top of the shared BaseIssue shape. |
| tests/unit/parsers/test_sarif_parser.py | Adds parser coverage for real fixtures and synthetic SARIF edge cases. |
| pyproject.toml | Registers the new SARIF parser package. |

</details>

<sub>Reviews (6): Last reviewed commit: ["fix(test): widen SARIF fixture helper ty..."](https://github.com/lgtm-hq/py-lintro/commit/27883bccb1eb6fb26b66994debf933352f720f26) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=42173607)</sub>

### CODE REVIEW

File: lintro/parsers/sarif/sarif_parser.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    locations = result.get("locations") or []
+    physical: dict[str, Any] = {}
+    if locations and isinstance(locations[0], dict):
+        physical = locations[0].get("physicalLocation") or {}
+    if not physical:
+        logger.debug("Skipping SARIF result without a physical location")
+        return None
+
+    artifact = physical.get("artifactLocation") or {}
+    file_path = _uri_to_path(str(artifact.get("uri", "")))
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/parsers/sarif/sarif_parser.py:180]: Addressed in 7d5d3ad1. The PoC deliberately does not resolve uriBaseId (tools lintro invokes from the project root emit workspace-relative URIs, which matches lintro's path model), but that limitation is now explicit: a code comment at the artifactLocation handling and a new 'URI base resolution' row in the evaluation's risk table, flagging that a production parser must resolve run.originalUriBaseIds (semgrep emits uriBaseId).

File: docs/design/sarif-ingestion-evaluation.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+  output (it re-runs the tool in fix mode), so this is not consumed.
+- **Per-tool enrichment.** `BanditIssue` composes a message from
+  `test_id`/`test_name`/`confidence`; `SemgrepIssue` carries `cwe`/`category`.
+  These live in SARIF under `result.properties` / `rule.properties`, but as
+  tool-specific bags — a shared parser can pass them through generically but
+  cannot reproduce each tool's bespoke message formatting without per-tool code,
+  which defeats the point.
+- **Doc URLs.** hadolint (verified) emits SARIF with **no** `rules[]` array, so
+  `helpUri` is unavailable and `doc_url` comes back empty — whereas lintro's
+  hadolint parser can synthesize a stable doc URL from a template. Doc-URL
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply docs/design/sarif-ingestion-evaluation.md]: Fixed in 7d5d3ad1: reworded to state that lintro's current hadolint parser does not populate doc_url either, and that the comparison is about capability — a native per-tool parser could synthesize a stable wiki URL from the rule code, whereas a generic SARIF parser has nothing to synthesize from when rules[] is absent.
